### ISSUE TITLE
How many times does getItemsCssClass() gets called

### ISSUE THREAD

**Issue by john20xdoe** [volkanceylan/Serenity#1238]
I was overriding `getItemCssClass()` method to apply the **conditional formatting** Basic Samples in Serene.

inside the method we place the validation rules (e.g. unitprice > 0, etc) before adding custom class. I added a console.log to check the contents `item` parameter (the first parameter of the method), which refers to the row.

in my code, it seems that the getItemCssClass() gets called several times per row (the same row data are repetitively appearing in the browser log, but some rows are not). Are there any docs or references that explains this? is the method called per-column per-row?.

Also, if you understand my purpose and I'm doing it wrong, is there an alternative method for my purpose (that traverses per row, so i can use its data to set the conditional formatting)? _(using getItemCssClass() seems wrong to me, since its for getting css?). I have scoured the source code for EntityGrid but can't find any._

**Comment by volkanceylan** [comment]
It is called by SlickGrid. You shouldn't worry about how many times. If performance is a concern, cache it in row itself, e.g. in a property __cssClass.

**Comment by john20xdoe** [comment]
Ok I'll try to save the css classes' string during the repository's ListHandler > ProcessEntity(MyRow row). 

but in SlickGrid, there's no alternative right? which is called per row?

Additional Note:
In the `item` parameter in getItemsCssClass(), I can also access the (hidden) fields I need to evaluate to conditional-format other columns (e.g. IsActive db column from some join) WHEN I include them in the XColumns. The downside, they also appear in the grid (shouldn't). When I set their attribute `Hidden` or `Visible(false)`, they disappear from `item` parameter. (So, I guess `item` refers to SlickGrid's columns, which are affected by XColumns, and not the XRow).

**Comment by volkanceylan** [comment]
You can add any arbitrary data to a row in client side, and it won't be shown anywhere. You shouldn't calculate that class server side, client side.

SlickGrid should call getItemCssClass once per row, but it might again due to reload etc.

**Comment by john20xdoe** [comment]
Um, by "client-side", i mean by typescript...
so I'm confused by this:
> You shouldn't calculate that class server side, client side.

Anyway, I ended up doing:
1. adding a [ClientSide, OneWay] string property *_ValidationCssClasses* for the row.
2. generating the css classes in MyListHandler > ProcessEntity(MyRow row) based on validation rules, and appending them to the row's *_ValidationCssClasses*. 
3. in the grid's getItemsCssClass(), I set the row css from the item's _ValidationCssClasses.
4. added the corresponding styles in the *.less* file.

It loaded way faster than when adding the validations solely in the grid's getItemsCssClass(). (My validations included executing a Q.getLookUp() for the row's joins (which reach to three levels), thus the lag.)

Also, the css are applied automatically when updating by dialog without needing to call any other method to trigger a refresh.
